packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelProviders/ModelProviderTests.cs

```diff
@@ -349,43 +349,58 @@ public void BuildModelAsStruct()
             var inputModel = InputFactory.Model("TestModel", properties: [], additionalProperties: additionalProperties);
             var modelTypeProvider = new ModelProvider(inputModel);
             var fields = modelTypeProvider.Fields;
 
-            // Assert
             Assert.IsNotNull(fields);
-            if (containsUnknownAdditionalProp)
+
+            if (containsMixedAdditionalProperties)
             {
-                Assert.AreEqual(0, fields.Count);
+                Assert.AreEqual(4, fields.Count);
+                Assert.AreEqual("_additionalBinaryDataProperties", fields[0].Name);
+                Assert.AreEqual(new CSharpType(typeof(IDictionary<string, BinaryData>)), fields[0].Type);
+                Assert.AreEqual("_additionalDoubleProperties", fields[1].Name);
+                Assert.AreEqual(new CSharpType(typeof(IDictionary<string, double>)), fields[1].Type);
+                Assert.AreEqual("_additionalInt64Properties", fields[2].Name);
+                Assert.AreEqual(new CSharpType(typeof(IDictionary<string, long>)), fields[2].Type);
+                Assert.AreEqual("_additionalStringProperties", fields[3].Name);
+                Assert.AreEqual(new CSharpType(typeof(IDictionary<string, string>)), fields[3].Type);
             }
             else
             {
                 Assert.AreEqual(1, fields.Count);
-                Assert.AreEqual("_serializedAdditionalRawData", fields[0].Name);
-                var type = fields[0].Type;
-                Assert.IsTrue(type.IsCollection);
+                Assert.AreEqual("_additionalBinaryDataProperties", fields[0].Name);
+                Assert.AreEqual(new CSharpType(typeof(IDictionary<string, BinaryData>)), fields[0].Type);
             }
         }
 
         [TestCaseSource(nameof(BuildAdditionalPropertiesTestCases))]
-        public void TestBuildAdditionalProperties(InputType additionalProperties, bool isUnverifiableType)
+        public void TestBuildAdditionalProperties(
+            InputType additionalProperties,
+            bool isUnverifiableType,
+            int expectedPropertyCount)
         {
             var inputModel = InputFactory.Model("TestModel", properties: [], additionalProperties: additionalProperties);
             var modelTypeProvider = new ModelProvider(inputModel);
 
-            var additionalPropertiesProp = modelTypeProvider.Properties.FirstOrDefault(f => f.Name == "AdditionalProperties");
-            Assert.IsNotNull(additionalPropertiesProp);
-
-            var additionalPropertiesType = additionalPropertiesProp!.Type;
-            Assert.IsTrue(additionalPropertiesType.IsDictionary);
+            var additionalPropertiesProps = modelTypeProvider.Properties.Where(f => f.Name.StartsWith("Additional")).ToList();
+            Assert.AreEqual(expectedPropertyCount, additionalPropertiesProps.Count);
 
-            if (isUnverifiableType)
+            foreach (var additionalPropertiesProp in additionalPropertiesProps)
```

Comment:
Do we have a test case for IReadOnlyDictionary?

Reply:
Added.